packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs

```diff
@@ -0,0 +1,226 @@
+use std::fmt;
+
+use error_stack::{Context, IntoReport, ResultExt};
+use regex::{Captures, Regex};
+use type_system::{DataType, EntityType, LinkType, PropertyType};
+
+#[derive(Debug)]
+pub struct DomainValidationError;
+
+impl Context for DomainValidationError {}
+
+impl fmt::Display for DomainValidationError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.write_str("URL failed to validate")
+    }
+}
+
+pub trait ValidateOntologyType<T> {
+    /// Checks a given type's ID against the given domain validation regex.
+    ///
+    /// # Errors
+    ///
+    /// - [`DomainValidationError`], if the base URI doesn't match or the kind is invalid
+    fn validate(&self, ontology_type: &T) -> error_stack::Result<(), DomainValidationError>;
+}
+
+#[expect(dead_code, reason = "We currently don't validate the shortname")]
+struct ShortNameAndKind<'a> {
+    pub short_name: &'a str,
+    pub kind: &'a str,
+}
+
+/// Responsible for validating Type URIs against a known valid pattern.
+#[derive(Clone)]
+pub struct DomainValidator(Regex);
+
+impl DomainValidator {
+    #[must_use]
+    /// Creates a new `DomainValidator`
+    ///
+    /// # Panics
+    ///
+    /// - If the "shortname" or "kind" named capture groups are missing
+    pub fn new(regex: Regex) -> Self {
+        regex
+            .capture_names()
+            .position(|capture_name| capture_name == Some("shortname"))
+            .expect("shortname capture group was missing");
+
+        regex
+            .capture_names()
+            .position(|capture_name| capture_name == Some("kind"))
+            .expect("kind capture group was missing");
+
+        Self(regex)
+    }
```

Comment:
Do we want to return an error on invalid regexes?

Reply:
Do you mean something that won't compile to Regex? Because that will already happen:
<img width="950" alt="image" src="https://user-images.githubusercontent.com/25749103/189124595-4e780495-a47a-451c-8bfb-56e17f75b5ba.png">

Also I stuck with panics because this isn't recoverable. We need a valid value for the domain validation URL

Reply:
This is an invalid regex and will fail when parsing the `Regex`. In this function you already have a valid regex, but it's not guaranteed that it fulfills our needs.

As this isn't recoverable and will not occur in runtime is good enough I guess, however, the error when passing an incomplete regex is not as nice as with `error-stack`:
```
$ cargo run -- --allowed-url-domain 'http://localhost:3000'

thread 'main' panicked at 'shortname capture group was missing', lib/graph/src/ontology/domain_validator.rs:41:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
This is only a nit and I don't want to block this PR on that, especially as we will revisit error handling very soon.

Reply:
Okay so you were trying to ask if we should return errors instead of `expect` for the missing named capture groups.
Considering we don't have a consolidated error strategy yet, and we haven't discussed how to handle unrecoverable errors, I'm not going to change this to a result and an error here. Because as discussed, it's not recoverable so it should be a panic. If we want to look at using error-stack for panics that's a separate discussion I think we should have outside of this thread.